src/insertItem.ts

```diff
@@ -152,18 +152,63 @@ export class InsertItem {
         };
         let defaultValue = await this.ui.showInputBox({ prompt: "Default value? Leave empty for no default value.", });
         if (defaultValue) {
-            parameter.defaultValue = defaultValue;
+            parameter.defaultValue = this.getDefaultValue(defaultValue, parameterType.value, context);
         }
         let description = await this.ui.showInputBox({ prompt: "Description? Leave empty for no description.", });
         if (description) {
             parameter.metadata = {
                 description: description
             };
         }
-        await this.insertInObject(template, textEditor, templateKeys.parameters, parameter, name, context);
+        await this.insertInObject(template, textEditor, templateKeys.parameters, parameter, name, context, false);
     }
 
-    private async insertInObject(template: DeploymentTemplate, textEditor: vscode.TextEditor, part: string, data: Data | unknown, name: string, context: IActionContext): Promise<void> {
+    private throwError(message: string, context: IActionContext, suppressReportIssue: boolean = true): void {
+        context.errorHandling.suppressReportIssue = true;
+        throw new Error(message);
+    }
+
+    private getDefaultValue(defaultValue: string, parameterType: string, context: IActionContext): string | number | boolean | undefined {
+        switch (parameterType) {
+            case "int":
+                let intValue = Number(defaultValue);
+                if (isNaN(intValue)) {
+                    this.throwError("Invalid integer!", context);
+                }
+                return intValue;
+            case "array":
+                try {
+                    // tslint:disable-next-line: no-unsafe-any
+                    return JSON.parse(defaultValue);
+                } catch (error) {
+                    this.throwError("Invalid array!", context);
+                }
+                break;
+            case "object":
+            case "secureobject":
+                try {
+                    // tslint:disable-next-line: no-unsafe-any
+                    return JSON.parse(defaultValue);
+                } catch (error) {
+                    this.throwError("Invalid object!", context);
+                }
+                break;
+            case "bool":
+                switch (defaultValue) {
+                    case "true":
+                        return true;
+                    case "false":
+                        return false;
+                    default:
+                        this.throwError("Invalid boolean!", context);
+                }
+                break;
+            default:
+                return defaultValue;
+        }
+    }
+
+    private async insertInObject(template: DeploymentTemplate, textEditor: vscode.TextEditor, part: string, data: Data | unknown, name: string, context: IActionContext, setCursor: boolean = true): Promise<void> {
```

Comment:
calls to insertInObject and insertInObjectHelper are becoming difficult to read, especially with that bool parameter.  Time to switch to using an options argument or else use the destructured object refactoring:
<img width="408" alt="image" src="https://user-images.githubusercontent.com/6913354/81018970-761b0e80-8e1a-11ea-89ca-66ad91442808.png">
It's then especially legible if everything's on its own line:
<img width="345" alt="image" src="https://user-images.githubusercontent.com/6913354/81019537-b62ec100-8e1b-11ea-83aa-3a1aef039171.png">
<img width="437" alt="image" src="https://user-images.githubusercontent.com/6913354/81019611-de1e2480-8e1b-11ea-946d-b04f418319e1.png">
(still a bit undecided on destructured object vs options object, IMHO options arg is easier to read)

Reply:
Switched to destructured object

Reply:
Fixed.